### PR TITLE
Update WikiaView to allow for finer grain template locations/names

### PR DESCRIPTION
**Pull request by garthwebb**
This allows controller methods to define a @template annotation which will determine what base template name will be used to render that method's output, e.g.:

``` php
class FooController {
  const DEFAULT_TEMPLATE_ENGINE = \WikiaResponse::TEMPLATE_ENGINE_MUSTACHE;

  /**
   * @template myView
   */
  public function bar() {
    // code
  }
}
```

The view will look for a series of possible template names, falling back until it finds one that exists.  So in the above example it will check:
- extensions/wikia/Foo/templates/myView.mustache
- extensions/wikia/Foo/templates/Foo_bar.mustache
- extensions/wikia/Foo/templates/FooController_bar.mustache

This pull request also allows the controller to define an alternate template path via a `getTemplateDir` method.  If this method exists, the return value will be used as the template directory:

``` php
class FooController {
  public static function getTemplateDir() {
    return dirname( __FILE__ ) . '/subfolder/templates';
  }
}
```

In this case the search paths will look like:
- extensions/wikia/Foo/subfolder/templates/myView.mustache
- extensions/wikia/Foo/subfolder/templates/Foo_bar.mustache
- extensions/wikia/Foo/subfolder/templates/FooController_bar.mustache

Note if the directory given does not exist, it will fall back to the default location.

@Wikia/platform-team

### CODE REVIEW

File: includes/wikia/nirvana/WikiaView.class.php [hunk 1]
```diff
@@ -31,16 +31,12 @@ function __construct( WikiaResponse $response = null ) {
 	 * @param string $methodName
 	 * @param array $data
 	 * @param string $format
+	 *
+	 * @return WikiaView
 	 */
-	public static function newFromControllerAndMethodName( $controllerName, $methodName, Array $data = array(), $format = WikiaResponse::FORMAT_HTML ) {
-		$app = F::app();
+	public static function newFromControllerAndMethodName( $controllerName, $methodName, Array $data = [], $format = WikiaResponse::FORMAT_HTML ) {
 		// Service classes must be dispatched by full name otherwise we default to a controller.
```
Comment (garthwebb) [inline includes/wikia/nirvana/WikiaView.class.php]: The following 6 lines were functionally duplicated later in code so replaced by common method ```normalizeControllerClass```

File: includes/wikia/nirvana/WikiaView.class.php [hunk 2]
```diff
@@ -114,30 +109,22 @@ protected function buildTemplatePath( $controllerClass, $methodName, $forceRebui
 			}
 
 			// Service classes must be dispatched by full name otherwise we default to a controller.
```
Comment (garthwebb) [inline includes/wikia/nirvana/WikiaView.class.php]: Lines from here up to L126 were rolled into common function ```normalizeControllerClass```

File: includes/wikia/nirvana/WikiaView.class.php [hunk 3]
```diff
@@ -114,30 +109,22 @@ protected function buildTemplatePath( $controllerClass, $methodName, $forceRebui
...
-				$controllerClass = $app->getServiceClassName( $controllerBaseName );
-			} else {
-				$controllerClass = $app->getControllerClassName( $controllerBaseName );
-			}
-
-			if( empty( $wgAutoloadClasses[$controllerClass] ) ) {
-				throw new WikiaException( "Invalid controller or service name: {$controllerClass}" );
-			}
-
-			// First we look for BaseName_MethodName
```
Comment (garthwebb) [inline includes/wikia/nirvana/WikiaView.class.php]: Put these changed lines into a loop looking at each fallback option.  The options are generated by ```getTemplateOptions```

**Comment by owend** [comment]
This LGTM.

File: includes/wikia/nirvana/WikiaView.class.php [hunk 4]
```diff
@@ -114,36 +109,106 @@ protected function buildTemplatePath( $controllerClass, $methodName, $forceRebui
...
+	 * @param $methodName
+	 *
+	 * @return array
+	 */
+	public function getTemplateOptions( $controllerClass, $methodName ) {
+		$templates = [];
+
+		// Make sure this method exists, otherwise the call to getMethod crashes PHP
+		// so badly it can't even log that a problem occurred.
+		if ( method_exists( $controllerClass, $methodName ) ) {
```
Comment (wladekb) [inline includes/wikia/nirvana/WikiaView.class.php]: I'd add an request-wide cache here for cases when you render many items separately so this doesn't add a lot of overhead (reflection used to be slow).

File: includes/wikia/nirvana/WikiaView.class.php [hunk 5]
```diff
@@ -114,36 +109,106 @@ protected function buildTemplatePath( $controllerClass, $methodName, $forceRebui
...
+
+	/**
+	 * See if the controller defines a custom template directory, otherwise use the default directory
+	 *
+	 * @param $controllerClass
+	 *
+	 * @return string
+	 */
+	public function getTemplateDir( $controllerClass ) {
+		if ( method_exists( $controllerClass, 'getTemplateDir' ) ) {
```
Comment (wladekb) [inline includes/wikia/nirvana/WikiaView.class.php]: I'd rather go with a dummy function returning null defined in base class so you can override it in a subclass if you need.

**Comment by wladekb** [comment]
left 2 comments, look good overall

**Comment by garthwebb** [comment]
@wladekb, can you take another look?  I've updated the PR

**Comment by garthwebb** [comment]
Ok, merging this PR after making the suggested changes.

**Pull request merged**